### DISCUSSION TITLE
useServerInsertedHTML twice in HTML code.

### DISCUSSION THREAD

**Discussion by dejurin** [vercel/next.js#49354] [Help] [answered]
### Summary

When I use **useServerInsertedHTML** I got 2 times in HTML code. 

<img width="1451" alt="Screenshot 2023-05-06 at 08 16 45" src="https://user-images.githubusercontent.com/4470015/236601740-b48e2cb1-431e-4c07-8b6c-ff920daa5789.png">

**styled.tsx**
```TS
'use client';
 
import { useServerInsertedHTML } from 'next/navigation';
 
export default function StyledComponentsRegistry({
  children,
}: {
  children: React.ReactNode;
}) {
 
  useServerInsertedHTML(() => {
    return <style>STYLE TWICE</style>;
  });
 
  if (typeof window !== 'undefined') return <>{children}</>;
 
  return (
    <>
      {children}
    </>
  );
}
```

**layout.tsx**
```TS
import './globals.css'
import { Inter } from 'next/font/google'
import StyledComponentsRegistry from './styled'
const inter = Inter({ subsets: ['latin'] })

export const metadata = {
  title: 'Create Next App',
  description: 'Generated by create next app',
}

export default function RootLayout({
  children,
}: {
  children: React.ReactNode
}) {
  return (
    <html lang="en">
      <body className={inter.className}><StyledComponentsRegistry>{children}</StyledComponentsRegistry></body>
    </html>
  )
}
```

### Additional information

```js
Apple M1 Max / OSX 13.3.1

"@types/node": "20.1.0",
"@types/react": "18.2.5",
"@types/react-dom": "18.2.4",
"eslint": "8.40.0",
"eslint-config-next": "13.4.1",
"next": "13.4.1",
"react": "18.2.0",
"react-dom": "18.2.0",
"typescript": "5.0.4"
```

**Comment by jyunhanlin** [comment, marked as answer]
Use `useRef` to make it injected once😆

```js
const isServerInserted = useRef(false);

useServerInsertedHTML(() => {
  if (!isServerInserted.current) {
    isServerInserted.current = true;
    const styles = styledComponentsStyleSheet.getStyleElement();
    styledComponentsStyleSheet.instance.clearTag();
    return <>{styles}</>;
  }
});
```

**Reply by dejurin** [reply]
Thank you, mate!

**Reply by gregoire-jianda** [reply]
Life savior

**Comment by Assem-Hafez** [comment]
Facing the same issue when trying to add hydrated css. Is this the expected behaviour ?

**Reply by itsmelion** [reply]
good question. smells like a bug to me.
Doesnt make sense to me that you have to do this at all.
its also undocumented on the examples.. it must at least be mentioned.
Hope it gets fixed ina future PR

**Comment by WilliamIPark** [comment]
I also wish there was some more information on this hook in the documentation. Right now it's only mentioned in regards to getting certain libraries working, but I have a use-case where based on some fetched data, I need to inject styles into the head. Tt seems like this hook might do that for me, but without documentation I'm not sure if it's supposed to be used in this way.
